src/python/pants/base/dependence_analysis.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from collections import defaultdict
+from dataclasses import dataclass
+from operator import itemgetter
+from typing import (
+    ClassVar,
+    DefaultDict,
+    Dict,
+    Generic,
+    Iterable,
+    List,
+    Optional,
+    Tuple,
+    Type,
+    TypeVar,
+    cast,
+)
+
+from pants.base.specs import Specs
+from pants.engine.goal import Goal
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.target import Sources, Target, Targets
+from pants.engine.unions import UnionMembership, union
+from pants.util.ordered_set import FrozenOrderedSet
+
+_S = TypeVar("_S", bound=Sources)
+GoalType = Type[Goal]
+
+
+@union
+@dataclass(frozen=True)
+class DependenceAnalysisRequest(Generic[_S]):
+    """A request to determine what kind of goal data dependence exists for target source fields."""
+
+    source_field_types: ClassVar[Tuple[Type[_S], ...]] = tuple()
```

Comment:
Not sure if it makes sense to be able to filter based on source field type. See tests for what it could look like.
Using field sets made even less sense, to me.